### ISSUE TITLE
MAE instead of L1 loss in pix2pix?

### ISSUE THREAD

**Issue by nobodykid** [eriklindernoren/Keras-GAN#150]
I noticed from your implementation of pix2pix, you use MAE as the second loss in [here](https://github.com/eriklindernoren/Keras-GAN/blob/master/pix2pix/pix2pix.py#L71)
```
self.combined.compile(loss=['mse', 'mae'],
                      loss_weights=[1, 100],
                      optimizer=optimizer)
```
But from the original implementation at [junyanz/pytorch-CycleGAN-and-pix2pix
](https://github.com/junyanz/pytorch-CycleGAN-and-pix2pix), they use L1 as the second loss. Is there a reason to your decision using MAE?

**Comment by CCInc** [comment]
Hi @nobodykid,

L1 loss and MAE are actually the same metrics. In the implementation you referenced, they use the `torch.nn.L1Loss` function, which in documentation is defined as:

> Creates a criterion that measures the mean absolute error (MAE) between each element in the input _x_ and the target _y_

Thus this implementation is correct for Keras.

**Comment by nobodykid** [comment]
Thank you very much. After inspecting `torch.nn.L1Loss` I found that there's an argument that helps you switching between MAE or L1 loss on Pytorch

**Comment by wave-transmitter** [comment]
Can anyone explain why loss weights are so imbalanced? According to the following: 

`self.combined.compile(loss=['mse', 'mae'],
                      loss_weights=[1, 100],
                      optimizer=optimizer)` 

it seems like MAE loss is multiplied by 100 combared to MSE.

**Comment by nobodykid** [comment]
It was according to the [paper](https://arxiv.org/abs/1611.07004). The 100 signifies the lambda coeff of MAE loss, which is the second term of the loss function used in pix2pix

**Comment by wave-transmitter** [comment]
@nobodykid thank you for your reply.

Still I think I am confused regarding the overall loss functions implementation. I can see that L1 loss is more or less the same to MAE. Yet, the adversarial loss is substituted in this implementation with MSE. Is this correct?

I have read most of the threads mentioning the misunderstanding regarding the selected loss functions, especially in #106, but I have not managed to understand why MSE is selected. Maybe @eriklindernoren can further help with that.

**Comment by DISAPPEARED13** [comment]
> @nobodykid thank you for your reply.
> 
> Still I think I am confused regarding the overall loss functions implementation. I can see that L1 loss is more or less the same to MAE. Yet, the adversarial loss of Pix2Pix defined as:
> 
> ![image](https://user-images.githubusercontent.com/26482319/80713800-b74eae00-8afc-11ea-9771-2782f9c28815.png)
> 
> is implemented with MSE. Is this correct? Should someone use binary cross-entropy instead?
> 
> I have read most of the threads mentioning the misunderstanding regarding the selected loss functions, especially in #106, but I have not managed to understand why MSE is selected. Maybe @eriklindernoren can further help with that.

I thought they define the function loss is log-like, and Mr.Zhu explained that loss optimizing by using MSE the residual of the variable. But in some gan hacks I saw everyone recommended BCE rather than MSE, the said that minimax problem is a 2-player problem so it should use binary cross entropy to calculate. if you just optimize G, MSE is better.
